File: python/binary-search/binary_search.py
def binary_search(list_of_numbers, number):
    if not list_of_numbers:
        raise ValueError("list_of_numbers is empty!")

    begin = 0
    end = len(list_of_numbers) - 1

    while begin <= end:
        middle = (begin + end) // 2
        if list_of_numbers[middle] == number:
            return middle
        if list_of_numbers[middle] < number:
            begin = middle + 1
        else:
            end = middle - 1

    raise ValueError("number is not in list_of_numbers!")

File: c/binary-search/src/binary_search.c
#include "binary_search.h"
#include <stdlib.h>

int* binary_search(int element, int* arr, unsigned int length) {
  if (!arr || length == 0 || element < arr[0] || element > arr[length - 1])
    return NULL;

  int* start = arr;
  int* end = arr + length - 1;
  int* middle;

  while (start <= end) {
    middle = start + (end - start) / 2;

    if (element < *middle)
      end = middle - 1;
    else if (element > *middle)
      start = middle + 1;
    else
      return middle;
  }

  return NULL;
}
